Feature/issue 1061

garyluu opened the pull request:
For issue ga4gh/dockstore#1061

Changes a few things from 'cwltool' to 'cwl-runner'.  Also adds a new section to the launch section of both tools and workflows based on what's specified in the issue.  
  
Also added a test

denis-yuen commented on src/app/workflow/version-modal/version-modal.component.ts:
```diff
@@ -124,4 +124,13 @@ export class VersionModalComponent implements OnInit, AfterViewChecked {
   // Validation ends here
+
+  // Checks if the currently edited test parameter file already exists
+  hasDuplicateTestJson() {
```
Isn't this just `return this.testParameterFilePaths.includes(this.testParameterFilePath)`?

garyluu replied:
Some weird merge caused this to appear.  I've rebased to remove irrelevant code

denis-yuen commented on src/app/workflow/launch/launch.component.html:
```diff
@@ -28,11 +28,16 @@
+      cwl-runner can run a tool directly when all inputs and outputs are available on the local filesystem.">
+        Run locally with cwl-runner
         <pre>{{ cwl }}</pre>
+        <!-- The below is disabled until workflows actually work -->
```
Worth pending a test for this on dockstore-webservice

garyluu replied:
The GA4GH production endpoint still works.  It might be a cwltool issue.

denis-yuen replied:
ok, create a new ticket for us to investigate and bookmark this line then so we follow up

garyluu replied:
Upon further investigation it looks like it does work for workflows.  It appears I was originally testing with a WDL workflow which explains why it didn't work.

denis-yuen commented on src/app/shared/launch.service.ts:
```diff
@@ -27,4 +27,10 @@ export abstract class LaunchService {
         return `$ consonance run --tool-dockstore-id ${path}:${versionName} ` +
             '--run-descriptor Dockstore.json --flavour \<AWS instance-type\>';
     }
+    getAltCwlString1(path: string, versionName: string) {
```
Let's make this more descriptive than "1" and "2"

denis-yuen commented on src/app/container/version-modal/version-modal.component.ts:
```diff
@@ -265,4 +266,23 @@ export class VersionModalComponent implements OnInit, AfterViewChecked {
+  // Checks if the currently edited test parameter file already exists
+  hasDuplicateTestJson(type) {
+    if (type === DescriptorType.CWL) {
+      if (this.unsavedCWLTestParameterFilePaths.includes(this.unsavedTestCWLFile)) {
```
This can be simpified

denis-yuen commented on src/app/container/launch/launch.component.html:
```diff
@@ -38,10 +38,14 @@
+        <a href="https://github.com/common-workflow-language/cwltool">Run locally with cwl-runner</a>
         <pre>{{ cwl }}</pre>
+        Alternatively, cwltool can run a tool off of dockstore.org when all inputs and outputs are available on the local filesystem or via HTTP(S):
+        <pre>{{ altCwl1 }}</pre>
```
Let's make the naming here more descriptive too

denis-yuen commented on src/app/container/launch/tool-launch.service.spec.ts:
```diff
@@ -41,14 +41,18 @@ describe('LaunchService', () => {
+    }));
+    it('should getAlternateStrings', inject([ToolLaunchService], (service: ToolLaunchService) => {
+        expect(service.getAltCwlString2('quay.io/briandoconnor/dockstore-tool-bamstats', '1.25-11')).toContain(
+            'cwltool --make-template quay.io/briandoconnor/dockstore-tool-bamstats:1.25-11 > input.yaml');
```
Is cwltool correct here?

garyluu replied:
The issue mentioning the new section shows cwltool as the example

denis-yuen commented on src/app/container/add-tag/add-tag.component.ts:
```diff
@@ -159,4 +167,23 @@ export class AddTagComponent implements OnInit, AfterViewChecked {
   }
   // Validation ends here
+  // Checks if the currently edited test parameter file already exists
+  // TODO: This code is repeated in version-modal.component.ts for tools, move it somewhere common
```
Yup

mr-c commented on src/app/container/launch/launch.component.html:
```diff
@@ -38,10 +38,14 @@
-      <div class="well well-sm" tooltip="Commands for launching tools through CWLtool: the CWL reference implementation.
-      cwltool can run a tool directly when all inputs and outputs are available on the local filesystem.">
-        <a href="https://github.com/common-workflow-language/cwltool">Run locally with cwltool</a>
+      <div class="well well-sm" tooltip="Commands for launching tools through cwl-runner: the CWL reference implementation.
```
`cwl-runner` is the generic way to run a CWL description using any CWL aware system. The name of the CWL reference runner is `cwltool`

mr-c replied:
so the second half of the sentence should read "the generic interface to many CWL implementations."

mr-c commented on src/app/container/launch/tool-launch.service.spec.ts:
```diff
@@ -41,14 +41,18 @@ describe('LaunchService', () => {
     it('should getCWLString', inject([ToolLaunchService], (service: ToolLaunchService) => {
         expect(service.getCwlString('quay.io/a/b', 'c'))
-            .toContain('cwltool --non-strict');
+            .toContain('cwl-runner --non-strict');
```
`--non-strict` shouldn't be here. It isn't part of the CWL standard and descriptions that need it are broken. I opened several PRs for prominent tools posted on dockstore.org to fix the "need" for `--non-strict" :-)

denis-yuen replied:
Thanks for the heads-up!
I think I'll have to take a look for those PRs, I may have missed the notifications. In the meantime, I agree that we cannot suggest `--non-strict` for cwl-runner if it only applies for cwltool. 

@garyluu  just note that `non-strict` or its equivalent for other engines can be a troubleshooting step whether via tooltip or shell comment

mr-c commented on src/app/shared/launch.service.ts:
```diff
@@ -27,4 +27,22 @@ export abstract class LaunchService {
+     * @param versionName The ToolVersion's name
+     */
+    getDockstoreSupportedCwlLaunchString(path: string, versionName: string) {
+        return `$ cwltool --non-strict ${path}:${versionName} Dockstore.json`;
```
I think this should be left out here. Users can add it back in if they wish

garyluu replied:
the '--non-strict'?

mr-c replied:
Correct, my apologies for the imprecision

garyluu commented:
Ready for re-review

denis-yuen commented on src/app/container/launch/launch.component.html:
```diff
@@ -38,10 +38,15 @@
-      cwltool can run a tool directly when all inputs and outputs are available on the local filesystem.">
-        <a href="https://github.com/common-workflow-language/cwltool">Run locally with cwltool</a>
+      <div class="well well-sm" tooltip="Commands for launching tools through cwl-runner when all inputs and outputs are available on the local filesystem.">
+        <a href="https://github.com/common-workflow-language/cwltool">Run locally with cwl-runner (you can use '--non-strict' when the executor is cwltool)</a>
```
Might want to explain why someone might want to do this and if it ends up being too long, maybe it can be hidden by a drop-down, info button, etc.

denis-yuen commented on src/app/container/launch/launch.component.html:
```diff
@@ -38,10 +38,15 @@
         <pre>{{ cwl }}</pre>
+        </div>
+        <div class="well well-sm" tooltip="Commands for launching tools through CWLtool: the CWL reference implementation. '--non-strict' can be used.">
+        Alternatively, cwltool can run a tool off of dockstore.org when all inputs and outputs are available on the local filesystem or via HTTP(S):
```
Same here, why would someone want to use non-strict. 
Also, what distinguishes this from the above option?

garyluu replied:
Still working on this...

denis-yuen commented on src/app/workflow/launch/launch.component.html:
```diff
@@ -28,12 +26,16 @@
-        Alternatively, cwltool can run a workflow directly when all inputs and outputs are available on the local filesystem
+    <div *ngIf="(currentDescriptor | lowercase) === 'cwl'">
+      <div class="well well-sm" tooltip="Commands for launching workflows through cwl-runner when all inputs and outputs are available on the local filesystem.">
+        Run locally with cwl-runner (you can use '--non-strict' when the executor is cwltool)
```
Ditto as above

denis-yuen commented:
Looks better, I think this just needs a bit more explanation, potentially hidden by default or linked to

garyluu commented:
Simplified a few things.  Moved the '--non-strict' and a brief explanation into tooltip to balance out the amount of content in tooltip vs description.  There is already a hyperlink to cwltool's GitHub.  People should honestly familiarize themselves with cwltool from there instead of Dockstore since there's quite a few flags.  Same applies for all other cwl-runners.
